### ISSUE TITLE
Seeking help for Level-8 accepting various formats 

### ISSUE THREAD

**Issue by ejnan** [nus-cs2103-AY2324S2/forum#243]
Hi everyone, I hope to get some help for level8 which I have been working on for some time. 

I am hoping the chatbot can accept dates and times of various formats but I keep getting different results whenever I add or change something. Everything worked when Deadline and Event was accepting String but I changed it to accept LocalDateTime and it does not work now.
And I've been staring at my code for too long so I am quite lost...

the inputs I hope it can accept are:
(year will be defaulted to current year if not specified. time is defaulted to end of the day, 23:59 if not specified)
```
deadline return book /by 2/12/2022 6:30 PM
deadline return book /by 2/12 6:30pm 
deadline return book /by 2 Dec 2022
deadline return book /by Dec 2 2022
deadline return book /by 2 Dec
deadline return book /by Dec 2
event project meeting /from 2 Dec 2022 2 PM /to 3 Dec 2022 4 PM
event project meeting /from Dec 2 2 PM /to Dec 3 4 PM
```
This is the part of my code I think is the issue - the part that parses the LocalDateTime
```
  private static final String[] PATTERNS = {
    "d/M/yyyy H:mm", "d/M/yyyy h:mm a", "d/M/yyyy h:mma", "d/M/yyyy h a", "d/M/yyyy ha",
    "d/M H:mm", "d/M h:mm a", "d/M h:mma", "d/M h a", "d/M ha",
    "d MMM yyyy H:mm", "d MMM yyyy h:mm a", "d MMM yyyy h:mma", "d MMM yyyy h a", "d MMM yyyy ha",
    "d MMM H:mm", "d MMM h:mm a", "d MMM h:mma", "d MMM h a", "d MMM ha",
    "MMM d yyyy H:mm", "MMM d yyyy h:mm a", "MMM d yyyy h:mma", "MMM d yyyy h a", "MMM d yyyy ha",
    "MMM d H:mm", "MMM d h:mm a", "MMM d h:mma", "MMM d h a", "MMM d ha",
    "d/M/yyyy", "d/M",
    "d MMM yyyy", "d MMM",
    "MMM d yyyy", "MMM d"
  };
```
this is the format I hope it to output, regardless of the input format
```
  [ D ][   ] deadline return book (by: Dec 2 2022 11:59 PM)
  [ D ][   ] return book (by: Dec 2 2022 6:30 PM)
  [ E ][   ] project meeting (from: Dec 2 2022 2:00 PM to: Dec 3 2022 4:00 PM)
```
Am I missing anything out that I have not noticed?
Thank you for your help in advance!

**Comment by kokerinks** [comment]
Hi! I did something similar for my Level-8 so maybe I can help. Is there anywhere I can refer to your current codebase? Are your most recent changes pushed to your branch-Level-8?

**Comment by ejnan** [comment]
I just [pushed it](https://github.com/ejnan/ip/tree/branch-Level-8/src/main/java)

**Comment by kokerinks** [comment]
1. Not too sure whether this will fix your issue, but one way you can clean up the patterns array is the split it into two different arrays - one array to store date formats and another array to store time formats. You can then use a nested for-loop and concatenate every single pairing of date and time formats while checking:

```
        String[] possibleDates = {
                "d/M/yyyy",
                "d-M-yyyy",
                "d/M/yy",
                "d-M-yy",
                "dMMyyyy",
                "dMMyy",

                "dd/MM/yyyy",
                "dd-MM-yyyy",
                "yyyy-MM-dd",

                "dd/MM/yy",
                "dd-MM-yy",
                "ddMMyyyy",
                "ddMMyy",};

        String[] possibleTimes = {"HHmm", "HH:mm", "HH", "h:mma",};

        for (String datePattern : possibleDates) {
            for (String timePattern : possibleTimes) {
                //check that time pattern comes before date pattern
                try {
                    DateTimeFormatter formatter = DateTimeFormatter.ofPattern(timePattern + " " + datePattern);
                    return LocalDateTime.parse(input, formatter);
                } catch (Exception e) {
                    //do nothing
                }
                //check that time pattern comes after date pattern
                try {
                    DateTimeFormatter formatter = DateTimeFormatter.ofPattern(datePattern + " " + timePattern);
                    return LocalDateTime.parse(input, formatter);
                } catch (Exception e) {
                    //do nothing
                }
            }
        }
```

The formats are clearer this way, and if it is an issue of an invalid datetime format hopefully you can catch it this way.

2. If the above doesn't fix it, maybe you can send the error which you are getting? Will give a better sensing of what is going wrong

**Comment by yisiox** [comment]
Hi @ejnan, you didn't specify what does not work so I'm assuming your question is why your code does not parse times involving AM and PM. If so, the issue originates in line 32 of your `DateTimeUtils`
```java
.parseDefaulting(ChronoField.HOUR_OF_DAY, 23) // Default to end of day (23:59)
```
`ChronoField.HOUR_OF_DAY` is for 24 hour time, so it conflicts with AM and PM. You can try the following to see that it's what's causing it to throw an error.
```java
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.time.format.DateTimeFormatterBuilder;
import java.time.temporal.ChronoField;
import java.util.Locale;

public class Test {
    public static void main(String[] args) {
        DateTimeFormatter formatter = new DateTimeFormatterBuilder()
                .appendPattern("d/M/yyyy h:mm a")
                .parseDefaulting(ChronoField.YEAR_OF_ERA, LocalDate.now().getYear())
                // .parseDefaulting(ChronoField.HOUR_OF_DAY, 23) // Default to end of day (23:59)
                // .parseDefaulting(ChronoField.MINUTE_OF_HOUR, 59)
                .parseDefaulting(ChronoField.CLOCK_HOUR_OF_AMPM, 11)
                .parseDefaulting(ChronoField.AMPM_OF_DAY, 1)
                .parseCaseInsensitive()
                .toFormatter(Locale.ENGLISH);
        String input = "2/12/2022 6:30 PM";
        LocalDateTime dateTime = LocalDateTime.parse(input, formatter);
        System.out.println(dateTime);
    }
}
```
That said I agree with @kokerinks that it's worth considering restructuring the code for handling date time (as the next increment after Level-8 involves major refactoring, so you can spare yourself some pain here).

**Comment by ejnan** [comment]
Thanks everyone for the help!

I did restructure the code as @kokerinks suggested - making it cleaner, and I think the issue @yisiox brought up was relevant too.

What else I did was: 
1. Make the parser flexible with spacing
2. Be able to assume same date as `/from` when it was not specified for `/to` for event 
3. Allow the parser to assume current year **and** end of day when both year and time were not specified for **only** deadline (my initial implementation only applied the defaults to when **either, not both,** was not specified)

What I wanted to achieve was flexibility of user input to make the task manager more user-friendly, but I did not expect there to be this many variations to be considered.
